# GEONOM

> Nomenclature of Countries and Territories
for the External Trade Statistics of the Community
and Statistics of Trade between Member States

GEONOM is published as a PDF document with tables giving the definitions of the alpha/numeric codes used to identify countries and territories.

In 2010, a correspondence table was published, defining the relationships between GEONOM identifiers and the ISO country codes. We'll use this for now.

In [1]:
import requests
from pathlib import Path
from io import BytesIO
from cachecontrol import CacheControl
from cachecontrol.caches.file_cache import FileCache
from cachecontrol.heuristics import LastModified
import pandas as pd

session = CacheControl(requests.Session(),
                       cache=FileCache('.cache'),
                       heuristic=LastModified())

inputURL = 'http://ec.europa.eu/eurostat/ramon/other_documents/geonom/concordances/geonom_2010-ISO.xls'
geonom = pd.read_excel(BytesIO(session.get(inputURL).content), na_values=[], keep_default_na=False)
geonom.drop(geonom.index[0], inplace=True)
geonom

Note ALPHA\nISO NUM\nISO  \
1                                                    *         AD       20   
2                                                    *         AE      784   
3                                                    *         AF        4   
4                                                    *         AG       28   
5                                                    *         AI      660   
6                                                    *         AL        8   
7                                                    *         AM       51   
8                                                    *         AN      530   
9                                                    *         AO       24   
10                                                   *         AQ       10   
11                                                   *         AR       32   
12                                                   *         AS       16   
13                                                   *         AT       40   
14                                                   *         AU       36   
15                                                   *         AW      533   
16                                                   *         AZ       31   
17                                                   *         BA       70   
18                                                   *         BB       52   
19                                                   *         BD       50   
20                                                   *         BE       56   
21                                                   *         BF      854   
22                                                   *         BG      100   
23                                                   *         BH       48   
24                                                   *         BI      108   
25                                                   *         BJ      204   
26                                                   *         BM       60   
27                                                   *         BN       96   
28                                                   *         BO       68   
29                                                   *         BR       76   
30                                                   *         BS       44   
..                                                 ...        ...      ...   
250                ISO 3166 codes included in EU codes                       
251                                                                          
252                                                 **         AX      248   
253                                                 **         BL      652   
254                                                 **         GF      254   
255                                                 **         GG      831   
256                                                 **         GP      312   
257                                                 **         IM      833   
258                                                 **         JE      832   
259                                                 **         MC      492   
260                                                 **         MF      663   
261                                                 **         MQ      474   
262                                                 **         PR      630   
263                                                 **         RE      638   
264                                                 **         SJ      744   
265                                                                          
266                                                                          
267                                                                          
268                                                                          
269                                                                     

The first part of the table is the main correspondence list, up to the line `ISO 3166 codes included in EU codes`. We'll just use these for now and remove the blank rows.

In [2]:
geonom = geonom[:geonom[geonom['Note'] == 'ISO 3166 codes included in EU codes'].index[0] -1]
geonom = geonom[geonom['COUNTRY'] != '']
geonom

Note ALPHA\nISO NUM\nISO ALPHA\nEU NUM \nEU  \
1      *         AD       20        AD       43   
2      *         AE      784        AE      647   
3      *         AF        4        AF      660   
4      *         AG       28        AG      459   
5      *         AI      660        AI      446   
6      *         AL        8        AL       70   
7      *         AM       51        AM       77   
8      *         AN      530        AN      478   
9      *         AO       24        AO      330   
10     *         AQ       10        AQ      891   
11     *         AR       32        AR      528   
12     *         AS       16        AS      830   
13     *         AT       40        AT       38   
14     *         AU       36        AU      800   
15     *         AW      533        AW      474   
16     *         AZ       31        AZ       78   
17     *         BA       70        BA       93   
18     *         BB       52        BB      469   
19     *         BD       50        BD      666   
20     *         BE       56        BE       17   
21     *         BF      854        BF      236   
22     *         BG      100        BG       68   
23     *         BH       48        BH      640   
24     *         BI      108        BI      328   
25     *         BJ      204        BJ      284   
26     *         BM       60        BM      413   
27     *         BN       96        BN      703   
28     *         BO       68        BO      516   
29     *         BR       76        BR      508   
30     *         BS       44        BS      453   
..   ...        ...      ...       ...      ...   
216    *         UA      804        UA       72   
217    *         UG      800        UG      350   
218    *         UM      581        UM      832   
219   **      US+PR  840+630        US      400   
220    *         UY      858        UY      524   
221    *         UZ      860        UZ       81   
222    *         VA      336        VA       45   
223    *         VC      670        VC      467   
224    *         VE      862        VE      484   
225    *         VG       92        VG      468   
226    *         VI      850        VI      457   
227    *         VN      704        VN      690   
228    *         VU      548        VU      816   
229    *         WF      876        WF      811   
230    *         WS      882        WS      819   
231    *         YE      887        YE      653   
232    *         YT      175        YT      377   
233    *         ZA      710        ZA      388   
234    *         ZM      894        ZM      378   
235    *         ZW      716        ZW      382   
238  >>>                            EU      999   
239  >>>                            QQ      950   
240  >>>                            QR      951   
241  >>>                            QS      952   
242  >>>                            QU      958   
243  >>>                            QV      959   
244  >>>                            QW      960   
246  >>>                            QX      977   
247  >>>                            QY      978   
248  >>>                            QZ      979   

                                               COUNTRY  \
1                                              Andorra   
2                                 United Arab Emirates   
3                                          Afghanistan   
4                                  Antigua and Barbuda   
5                                             Anguilla   
6                                              Albania   
7                                              Armenia   
8                                 Netherlands Antilles   
9                                               Angola   
10                                          Antarctica   
11                                           Argentina   
12                                      American Samoa   
13                                             Austria   
14                           

Note that Pandas has interpreted Excel cells as numbers and so has lost the leading 0s in columns `NUM ISO` and `NUM EU`. These are three digit codes, although being a correspondence value, the `NUM ISO` column lists some codes differently, e.g. `246 + 248`.

Before we fix that, the column headers have picked up newlines and spaces, so rename.

In [3]:
geonom.columns

Index(['Note', 'ALPHA\nISO', 'NUM\nISO', 'ALPHA\nEU', 'NUM \nEU', 'COUNTRY',
       'DESCRIPTION'],
      dtype='object')

In [4]:
geonom.rename(columns={
    'ALPHA\nISO': 'ALPHA_ISO', 'NUM\nISO': 'NUM_ISO',
    'ALPHA\nEU': 'ALPHA_EU', 'NUM \nEU': 'NUM_EU'}, inplace=True)
geonom.columns

Index(['Note', 'ALPHA_ISO', 'NUM_ISO', 'ALPHA_EU', 'NUM_EU', 'COUNTRY',
       'DESCRIPTION'],
      dtype='object')

Ensure leading zeros are in `NUM_EU`

In [5]:
geonom['NUM_EU'] = geonom['NUM_EU'].apply(lambda x: "%03d" % int(x))

In [6]:
geonom

Note ALPHA_ISO  NUM_ISO ALPHA_EU NUM_EU  \
1      *        AD       20       AD    043   
2      *        AE      784       AE    647   
3      *        AF        4       AF    660   
4      *        AG       28       AG    459   
5      *        AI      660       AI    446   
6      *        AL        8       AL    070   
7      *        AM       51       AM    077   
8      *        AN      530       AN    478   
9      *        AO       24       AO    330   
10     *        AQ       10       AQ    891   
11     *        AR       32       AR    528   
12     *        AS       16       AS    830   
13     *        AT       40       AT    038   
14     *        AU       36       AU    800   
15     *        AW      533       AW    474   
16     *        AZ       31       AZ    078   
17     *        BA       70       BA    093   
18     *        BB       52       BB    469   
19     *        BD       50       BD    666   
20     *        BE       56       BE    017   
21     *        BF      854       BF    236   
22     *        BG      100       BG    068   
23     *        BH       48       BH    640   
24     *        BI      108       BI    328   
25     *        BJ      204       BJ    284   
26     *        BM       60       BM    413   
27     *        BN       96       BN    703   
28     *        BO       68       BO    516   
29     *        BR       76       BR    508   
30     *        BS       44       BS    453   
..   ...       ...      ...      ...    ...   
216    *        UA      804       UA    072   
217    *        UG      800       UG    350   
218    *        UM      581       UM    832   
219   **     US+PR  840+630       US    400   
220    *        UY      858       UY    524   
221    *        UZ      860       UZ    081   
222    *        VA      336       VA    045   
223    *        VC      670       VC    467   
224    *        VE      862       VE    484   
225    *        VG       92       VG    468   
226    *        VI      850       VI    457   
227    *        VN      704       VN    690   
228    *        VU      548       VU    816   
229    *        WF      876       WF    811   
230    *        WS      882       WS    819   
231    *        YE      887       YE    653   
232    *        YT      175       YT    377   
233    *        ZA      710       ZA    388   
234    *        ZM      894       ZM    378   
235    *        ZW      716       ZW    382   
238  >>>                          EU    999   
239  >>>                          QQ    950   
240  >>>                          QR    951   
241  >>>                          QS    952   
242  >>>                          QU    958   
243  >>>                          QV    959   
244  >>>                          QW    960   
246  >>>                          QX    977   
247  >>>                          QY    978   
248  >>>                          QZ    979   

                                               COUNTRY  \
1                                              Andorra   
2                                 United Arab Emirates   
3                                          Afghanistan   
4                                  Antigua and Barbuda   
5                                             Anguilla   
6                                              Albania   
7                                              Armenia   
8                                 Netherlands Antilles   
9                                               Angola   
10                                          Antarctica   
11                                           Argentina   
12                                      American Samoa   
13                                             Austria   
14                                           Australia   
15                                               Aruba   
16                                          Azerbaijan   
17                              Bosnia and Herzegovina   
18                                           

In [7]:
from rdflib import Graph, Literal, BNode, Namespace, RDF, URIRef, RDFS, OWL, XSD
from rdflib.namespace import SKOS
from rdflib.collection import Collection
import numpy as np

GN = Namespace('http://gss-data.org.uk/def/geonom_2012#')

g = Graph()
g.bind('gn', GN)
g.bind('skos', SKOS)

scheme = URIRef('http://gss-data.org.uk/def/geonom_2012')
g.add((scheme, RDF.type, SKOS.ConceptScheme))
g.add((scheme, RDFS.label, Literal('Geonomenclature, 2010')))

for dt in [{'subj': GN.Alpha2,
            'label': 'Alpha 2',
            'comment': 'Two letter country code',
            'pattern': '[A-Z]{2}'
           },
           {'subj': GN.Num3,
            'label': 'Numeric 3',
            'comment': 'Numeric three digit code',
            'pattern': '[0-9]{3}'
           }
          ]:
    g.add((dt['subj'], RDF.type, RDFS.Datatype))
    g.add((dt['subj'], RDFS.label, Literal(dt['label'])))
    g.add((dt['subj'], RDFS.comment, Literal(dt['comment'])))
    g.add((dt['subj'], OWL.onDatatype, XSD.String))
    restriction_resource = BNode()
    restrictions = Collection(g, restriction_resource)
    g.add((dt['subj'], OWL.withRestrictions, restriction_resource))
    pattern_resource = BNode()
    restrictions.append(pattern_resource)
    g.add((pattern_resource, XSD.pattern, Literal(dt['pattern'])))

for i, row in geonom.iterrows():
    term = GN.term(row['ALPHA_EU'])
    g.add((term, RDF.type, SKOS.Concept))
    g.add((term, RDFS.label, Literal(row['COUNTRY'].strip())))
    g.add((term, SKOS.inScheme, scheme))
    if row['DESCRIPTION'].strip() != '':
        g.add((term, RDFS.comment, Literal(row['DESCRIPTION'].strip())))
    g.add((term, SKOS.notation, Literal(row['ALPHA_EU'], datatype=GN.Alpha2)))
    g.add((term, SKOS.notation, Literal(row['NUM_EU'], datatype=GN.Alpha2)))
    
print(g.serialize(format='n3').decode('utf-8'))

@prefix gn: <http://gss-data.org.uk/def/geonom_2012#> .
@prefix ns1: <http://www.w3.org/2002/07/owl#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix skos: <http://www.w3.org/2004/02/skos/core#> .
@prefix xml: <http://www.w3.org/XML/1998/namespace> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

gn:AD a skos:Concept ;
    rdfs:label "Andorra" ;
    skos:inScheme <http://gss-data.org.uk/def/geonom_2012> .

gn:AE a skos:Concept ;
    rdfs:label "United Arab Emirates" ;
    rdfs:comment "Abu Dhabi, Dubai, Sharjah, Ajman, Umm al Qaiwain, Ras al Khaima and Fujairah" ;
    skos:inScheme <http://gss-data.org.uk/def/geonom_2012> .

gn:AF a skos:Concept ;
    rdfs:label "Afghanistan" ;
    skos:inScheme <http://gss-data.org.uk/def/geonom_2012> .

gn:AG a skos:Concept ;
    rdfs:label "Antigua and Barbuda" ;
    skos:inScheme <http://gss-data.org.uk/def/geonom_2012> .

gn:AI a skos:Concept ;
    rdfs:label "Angu

In [8]:
out = Path('out')
out.mkdir(exist_ok=True, parents=True)
with open(out / 'geonom_2010.ttl', 'wb') as f:
    g.serialize(f, format='n3')